processor/src/decoder/aux_trace/block_hash_table.rs

```diff
@@ -7,197 +9,261 @@ use super::{
+
+                // Note: this adds the 2 rows separately to the block hash table.
+                left_child_row.collapse(alphas) * right_child_row.collapse(alphas)
+            }
+            SPLIT => BlockHashTableRow::from_split(main_trace, row).collapse(alphas),
+            LOOP => BlockHashTableRow::from_loop(main_trace, row)
+                .map(|row| row.collapse(alphas))
+                .unwrap_or(E::ONE),
+            REPEAT => BlockHashTableRow::from_repeat(main_trace, row).collapse(alphas),
+            DYN => BlockHashTableRow::from_dyn(main_trace, row).collapse(alphas),
             _ => E::ONE,
         }
     }
 }
 
-// HELPER FUNCTIONS
+// BLOCK HASH TABLE ROW
 // ================================================================================================
 
-/// Computes the multiplicand representing the removal of a row from the block hash table.
-fn get_block_hash_table_removal_multiplicand<E: FieldElement<BaseField = Felt>>(
-    main_trace: &MainTrace,
-    i: usize,
-    alphas: &[E],
-    op_code_next: u8,
-) -> E {
-    let a = main_trace.addr(i + 1);
-    let digest = main_trace.decoder_hasher_state_first_half(i);
-    let is_loop_body = main_trace.is_loop_body_flag(i);
-    let next_end_or_repeat =
-        if op_code_next == END || op_code_next == REPEAT || op_code_next == HALT {
-            E::ZERO
+/// Describes a single entry in the block hash table. An entry in the block hash table is a tuple
+/// (parent_id, block_hash, is_first_child, is_loop_body), where each column is defined as follows:
+/// - parent_block_id: contains the ID of the current block. Note: the current block's ID is the
+///   parent block's ID from the perspective of the block being added to the table.
+/// - block_hash: these 4 columns hold the hash of the current block's child which will be executed
+///   at some point in time in the future.
+/// - is_first_child: set to true if the table row being added represents the first child of the
+///   current block. If the current block has only one child, set to false.
+/// - is_loop_body: Set to true when the current block block is a LOOP code block (and hence, the
+///   current block's child being added to the table is the body of a loop).
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct BlockHashTableRow {
+    parent_block_id: Felt,
+    child_block_hash: Word,
+    is_first_child: bool,
+    is_loop_body: bool,
+}
+
+impl BlockHashTableRow {
+    // CONSTRUCTORS
+    // ----------------------------------------------------------------------------------------------
+
+    // Computes the initial row in the block hash table.
+    pub fn table_init(main_trace: &MainTrace) -> Self {
+        let program_hash = {
+            let row_with_halt = (0..main_trace.num_rows())
+                .find(|row| main_trace.get_op_code(*row) == Felt::from(HALT))
+                .expect("execution trace must include at least one occurrence of HALT");
```

Comment:
Probably not for this PR, but I wonder how expensive this `.find()` is. Basically, we need to iterate through the whole trace to find the first `HALT` row.

One alternative to this is to get the program hash from the last row (excluding the random rows at the end). This should be possible because because the semantics of `HALT` are such that it copies over the entire state of the VM (including the decoder registers) to the next cycle. So, program hash should be the same in the last row as in the first `HALT` row.

Reply:
Created #1344